segregation/inference/comparative.py

```diff
@@ -0,0 +1,453 @@
+    data2 : pd.DataFrame or gpd.DataFrame
+        Pandas or Geopandas dataframe holding data for context 2
+
+    group_pop_var : str
+        The name of variable in both data that contains the population size of the group of interest
+
+    total_pop_var : str
+        The name of variable in both data that contains the total population of the unit
+
+    approach : str, ["composition", "share", "dual_composition"]
+        Which approach to use for generating the counterfactual.
+        Options include "composition", "share", or "dual_composition"
+
+    Returns
+    -------
+    two DataFrames
+        df1 and df2  with appended columns 'counterfactual_group_pop', 'counterfactual_total_pop', 'group_composition' and 'counterfactual_composition'
+
+    """
+    df1, df2 = DUAL_SIMULATORS[counterfactual_approach](
+        data1, data2, group_pop_var1, total_pop_var1, group_pop_var2, total_pop_var2,
+    )
+    df1["group_composition"] = (df1[group_pop_var1] / df1[total_pop_var1]).fillna(0)
+    df2["group_composition"] = (df2[group_pop_var2] / df2[total_pop_var2]).fillna(0)
+
+    df1["counterfactual_composition"] = (
+        df1["counterfactual_group_pop"] / df1["counterfactual_total_pop"]
+    ).fillna(0)
+    df2["counterfactual_composition"] = (
+        df2["counterfactual_group_pop"] / df2["counterfactual_total_pop"]
+    ).fillna(0)
+
+    df1 = df1.drop(columns=[group_pop_var1, total_pop_var1], axis=1)
+    df2 = df2.drop(columns=[group_pop_var2, total_pop_var2], axis=1)
+
+    return df1, df2
+
+
+def sim_composition(
+    df1, df2, group_pop_var1, total_pop_var1, group_pop_var2, total_pop_var2,
+):
+    """Simulate the spatial distribution of a population group in a region using the CDF of a comparison region.
+
+    For each spatial unit i in region 1, take the unit's rank in the distribution, and swap the group composition
+    with the value of the corresponding rank in region 2. The composition is the minority population of unit i
+    divided by total population of tract i. This approach will shift the relative composition of each spatial
+    unit without changing its total population.
+
+    Parameters
+    ----------
+    df1 : pandas.DataFrame or geopandas.GeoDataFrame
+        dataframe for first dataset with columns holding group and total population counts
+    df2 : pandas.DataFrame or geopandas.GeoDataFrame
+        dataframe for second dataset with columns holding group and total population counts
+    group_pop_var1 : str
+        column holding population counts for group of interest on input df1
+    total_pop_var1 : str
+        column holding total population counts on input df1
+    group_pop_var2 : str
+        column holding population counts for group of interest on input df2
+    total_pop_var2 : str
+        column holding total population counts on input df2
+
+    Returns
+    -------
+    two pandas.DataFrame
+        dataframes with simulated population columns appended
+    """
+    df1 = df1.copy()
```

Comment:
What's the reason for creating an identically named data variable? (and below)

Reply:
this one im not 100% sure is necessary, but think i'd prefer to be safe and leave it